src/scriptdev2/scripts/kalimdor/the_barrens.cpp

```diff
@@ -591,6 +592,123 @@ CreatureAI* GetAI_npc_wizzlecranks_shredder(Creature* pCreature)
     return new npc_wizzlecranks_shredderAI(pCreature);
 }
 
+/*#####
+## Quest Counterattack!
+QuestID 4,021
+#####*/
+enum
+{
+
+	NPC_KROMZAR = 9456,
+	NPC_KOLKAR_INVADER = 9524,
+	NPC_KOLKAR_STORMSEER = 9523,
+	NPC_HORDE_DEFENDER = 9457,
+	NPC_HORDE_AXE_THROWER = 9458,
+	NPC_REGTHAR_DEATHGATE = 3389,
+	QUEST_COUNTERATTACK = 4021,
+
+};
+
+struct npc_regthar_deathgateAI : public ScriptedAI {
+	npc_regthar_deathgateAI(Creature* pCreature) : ScriptedAI(pCreature) { Reset(); }
+
+
+
+
+	uint8 m_uiMobCount;
+
+	ObjectGuid m_playerGuid;
+
+	void Reset() override
+	{
+		m_playerGuid.Clear();
+	}
+
+	void JustDied(Unit* /*pKiller*/) override
+	{
+		Player* pPlayer = m_creature->GetMap()->GetPlayer(m_playerGuid);
+
+		if (pPlayer && pPlayer->GetQuestStatus(QUEST_COUNTERATTACK) == QUEST_STATUS_INCOMPLETE)
+			pPlayer->SendQuestFailed(QUEST_COUNTERATTACK);
+	}
+
+	void JustSummoned(Creature* pSummoned) override
+	{
+		pSummoned->AI()->AttackStart(m_creature);
+
+	}
+
+
+	void SummonedCreatureJustDied(Creature* /*pKilled*/) override
+	{
+		++m_uiMobCount;
+		if (m_uiMobCount == 10)
+			m_creature->SummonCreature(NPC_KROMZAR, -298.0f, -1836.0f, 94.06f, 0.0f, TEMPSUMMON_TIMED_OOC_OR_DEAD_DESPAWN, 900000); ///Summon in the middle of the invisible box
+
+		if (m_uiMobCount % 4 == 0) //for every 4 units kiled, bring back up
+			m_creature->SummonCreature(NPC_HORDE_DEFENDER, -290.512054f, -1872.060547f, 92.658226f, 0.0f, TEMPSUMMON_TIMED_OOC_OR_DEAD_DESPAWN, 600000);
+		if (m_uiMobCount)
+			return;
+
+
+		Player* pPlayer = m_creature->GetMap()->GetPlayer(m_playerGuid);
+
+		if (pPlayer && pPlayer->GetQuestStatus(QUEST_COUNTERATTACK) == QUEST_STATUS_INCOMPLETE)
+			pPlayer->GroupEventHappens(QUEST_COUNTERATTACK, m_creature);
+
+		m_playerGuid.Clear();
+		m_creature->SetFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_QUESTGIVER);
+	}
+
+
+	void StartEvent(Player* pPlayer)
```

Comment:
Remove quest giver flag, otherwise people will be able to start the event multiple times and bugging it.